spock-core/src/main/java/org/spockframework/runtime/SafeIterationNameProvider.java

```diff
@@ -14,31 +14,26 @@
 
 package org.spockframework.runtime;
 
-import java.util.concurrent.atomic.AtomicInteger;
-
 import org.spockframework.runtime.model.*;
 
 public class SafeIterationNameProvider implements NameProvider<IterationInfo> {
+  private final NameProvider<IterationInfo> fallback = new DataVariablesIterationNameProvider();
   private final NameProvider<IterationInfo> delegate;
-  private AtomicInteger iterationCount = new AtomicInteger();
 
   public SafeIterationNameProvider(NameProvider<IterationInfo> delegate) {
     this.delegate = delegate;
   }
 
   @Override
   public String getName(IterationInfo iteration) {
-    String safeName = iteration.getParent().isReportIterations() ?
-        String.format("%s[%d]", iteration.getParent().getName(), iterationCount.getAndIncrement()) : iteration.getParent().getName();
-
-    if (delegate == null) return safeName;
+    if (delegate == null) return fallback.getName(iteration);
```

Comment:
I'd still like to keep the previous implementation here, as it is a safe variant even if there is an exception with the other name provider. Just set the `DataVariablesIterationNameProvider` as an alternative to the `UnrollNameProvider`

Reply:
I'm not sure what you mean by `set as an alternative`.
Actually with above change, the new provider should be safe, but what do you think about this as last resort:
```groovy
} catch (Exception e) {
  try {
    return fallback.getName(iteration);
  } catch (Exception e2) {
    return format("%s [#%d]", iteration.getFeature().getName(), iteration.getIterationIndex());
  }
}
```
?

Reply:
I mean `org/spockframework/runtime/extension/builtin/UnrollExtension.java:50` instead of returning `null` here you would be returning the new `DataVariablesIterationNameProvider`, since we need to change the extension anyway to make unrolling and reporting work correctly

Reply:
I don't think `UnrollExtension` is the proper place.
Any other custom extension could set a custom iteration name provider.
Like it is now, this one is always used as default and fallback, which probably also was the case why the old logic was in `SafeIterationNameProvider` and not in `UnrollExtension`.

Reply:
Yes, and I would still like to have the old dead simple `#featureName[#iterationIndex]` logic in the `SafeIterationNameProvider` as fallback, and the `UnrollExtension` would just use `DataVariablesIterationNameProvider` as default if not other pattern is given.

Reply:
But that would also mean, that if `UnrollIterationNameProvider` returns `null` or throws an exception it just uses the "dead simple" variant. :-(

Reply:
yes, and in most cases where `UnrollIterationNameProvider` would fail `DataVariablesIterationNameProvider` would probably fail too